proposals/0002-testing-log-handler.md

```diff
@@ -0,0 +1,112 @@
+# Test Log Handler
+
+Authors: [Brian Maher](https://github.com/brimworks), [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+
+## Introduction
+
+Software operators rely on logging to provide insight into what is being done,
+therefore, testing log output is an important facet.
+
+## Motivation
+
+A variety of implementations have cropped up organically, but having a single
+test log handler to build off of will reduce the re-write churn.
+
+### Existing Implementations
+
+* [swift-log-testing](https://github.com/neallester/swift-log-testing)
+* [swift-distributed-actors/LogCapture](https://github.com/apple/swift-distributed-actors/blob/main/Sources/DistributedActorsTestKit/LogCapture.swift)
+* [swift-cluster-membership/LogCapture](https://github.com/apple/swift-cluster-membership/blob/main/Tests/SWIMTestKit/LogCapture.swift)
+* [RediStack/RedisLoggingTests](https://github.com/swift-server/RediStack/blob/master/Tests/RediStackIntegrationTests/RedisLoggingTests.swift)
+* [swift-log/TestLogger](https://github.com/apple/swift-log/blob/main/Tests/LoggingTests/TestLogger.swift#L52)
+
+
+## Features
+
+A testing log handler needs the following features:
+* The ability to capture log entries at the start of a unit test.
+* The ability to assert the captured log entries contain a matching message.
+* Work correctly in the face of parallel test execution.
+* Print log messages to the console if the log entry does not match a log entry
+  that is being asserted.
+* Bootstrap the test log handler as early as possible so we don't miss any
+  `Logger()` creations which may happen during setup.
+
+## Proposed Solution
+
+A global function to bootstrap the test log handler:
+
+```swift
+TestingLogHandler.bootstrap()
```

Comment:
Why don't we call the regular bootstrap method here? Furthermore, I think when we do this we should also make the init of `Logger` finally public that just takes a `LogHandler` so APIs that accept `Logger`s can be tested by just creating a `TestingLogHandler`, creating a logger with it and then calling the API with that logger.

Reply:
> Why don't we call the regular bootstrap method here?

As I recall, the `LoggingSystem.bootstrap()` must be called exactly once, so my thought was to encapsulate the logic for ensuring it is initialized exactly once into the `LoggingSystem.bootstrap()` method, but perhaps I am mistaken in my understanding?

> Furthermore, I think when we do this we should also make the init of Logger finally public that just takes a LogHandler so APIs that accept Loggers can be tested by just creating a TestingLogHandler, creating a logger with it and then calling the API with that logger.

I'm not opposed to that change, however I have found that it isn't to uncommon to find code which does not expose a way to "inject" the `Logger` instance (or trying to "inject" it with our custom log handler is rather onerous).

Overall, I do like the idea of making the `Logger` constructor that takes a `LogHandler` instance as public, but I'd like to also have a way to capture logs too.

Reply:
Honestly, I feel this situation needs to be resolved in a separate proposal where the bootstrap method is relaxed, perhaps only in debug mode.

Reply:
BTW, @FranzBusch when putting together the sample implementation I noticed that there already is `Logger()` constructors (aka `init()`) that allow one to specify a log factory, so it looks like this comment is already done:

>  Furthermore, I think when we do this we should also make the init of Logger finally public that just takes a LogHandler so APIs that accept Loggers can be tested by just creating a TestingLogHandler, creating a logger with it and then calling the API with that logger.

@Mordil  said:

> Honestly, I feel this situation needs to be resolved in a separate proposal where the bootstrap method is relaxed, perhaps only in debug mode.

Perhaps... but I do think that if we have a reasonably good test logger which allows one to do initialization exactly once, then there is no need to allow for multiple initialization.

Actually, one way to get around the multiple initialization problem is to specify a log factory that delegates to an implementation which can be changed... actually, maybe I should use that idea in the current log tests since I'm trying to allow the current tests to be minimally modified (which means supporting multiple log handlers).

Reply:
> Actually, one way to get around the multiple initialization problem is to specify a log factory that delegates to an implementation which can be changed... actually, maybe I should use that idea in the current log tests since I'm trying to allow the current tests to be minimally modified (which means supporting multiple log handlers).

So, I went ahead and made an attempt at writing a log factory delegate here:

https://github.com/brimworks/swift-log/blob/testing-log-handler-proposal-delegated/Sources/Logging/DelegatedLogHandler.swift

Unfortunately, the current log unit tests rely on the copy on write behavior of structs, but my `DelegatedLogHandler` needs to be a class so we can keep track of any outstanding `LogHandler`s and update them (using a weak map). If I can find a way to replicate the copy on write behavior of structs in a class then I think I could get this to pass the current tests.

Reply:
I am still not convinced this is the right thing to do in tests here. IMO if developers want to test their logs I don't think they should use bootstrapping of a test log handler but rather they should accept a `Logger` and we can create a `Logger(handler: TestLogerHandler()`. This would make this proposal a lot simpler since we don't need containers anymore.

Furthermore, it would push developers into the right pattern which is to accept a `Logger` at their APIs and potentially fallback to create a `Logger` if none is passed.